[AST] Generalize and rename ClassDecl::getEmittedMembers()

DougGregor opened the pull request:
Generalize `ClassDecl::getEmittedMembers()` to operate on an
`IterableDeclContext`, so that it can be for other nominal types,
extensions, etc. Rename to `getSemanticMembers()` to indicate that
these are all of the members that are semantically part of that
context.

Also add `getParsedMembers()` to retrieve the set of members
that were parsed from the source code, excluding any that were
implicitly synthesized by the implementation. Clients should choose
between these new entry points, because `getMembers()` itself
changes meaning as the translation progresses.

Clean up the implementation slightly so it only forces type checking
for the conformances within that particular context (using
`getLocalConformances()`) and doesn't need to list out each of the
protocols it cares about.

DougGregor committed:
[AST] Generalize and rename ClassDecl::getEmittedMembers()

Generalize `ClassDecl::getEmittedMembers()` to operate on an
`IterableDeclContext`, so that it can be for other nominal types,
extensions, etc. Rename to `getSemanticMembers()` to indicate that
these are all of the members that are semantically part of that
context.

Clean up the implementation slightly so it only forces type checking
for the conformances within that particular context (using
`getLocalConformances()`) and doesn't need to list out each of the
protocols it cares about.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

slavapestov commented on lib/AST/DeclContext.cpp:
```diff
@@ -793,6 +793,14 @@ DeclRange IterableDeclContext::getMembers() const {
+  ASTContext &ctx = getASTContext();
+  return evaluateOrDefault(
+      ctx.evaluator,
+      EmittedMembersRequest{const_cast<IterableDeclContext *>(this)},
```
Rename the request itself too then?

DougGregor replied:
Oh, duh

slavapestov commented on test/Incremental/Verifier/single-file-private/AnyObject.swift:
```diff
@@ -25,7 +25,6 @@ import Foundation
 // expected-private-conformance {{Swift.CVarArg}}
 // expected-private-conformance {{Swift.CustomStringConvertible}}
 // expected-private-member {{Swift._ExpressibleByBuiltinIntegerLiteral.init}}
-// expected-private-superclass {{main.LookupFactory}}
```
Is this expected?

DougGregor replied:
Yeah, we're no longer doing conformance lookup for a specific set of conformances. Rather, we ask for which conformances are ascribed to this particular decl context, and force those if they are derivable.

CodaFi replied:
I'm surprised to see these edges go, but it shouldn't affect anything in practice. The cross-file tests losing this dependency would be the catastrophic one.

slavapestov reviewed (approved):
Thank you for tackling this!

CodaFi commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2507,59 +2507,51 @@ struct SortedFuncList {
-        conformance->getState() == ProtocolConformanceState::Incomplete) {
+  // Force any derivable conformances in this context. This ensures that any
+  // synthesized members will approach in the member list.
+  for (auto conformance : idc->getLocalConformances()) {
```
This is a behavior change. Is it okay to ignore conformances from other decl contexts e.g. extensions?

slavapestov replied:
Yeah, because they add members to the extension so we'll synthesize them when you call getSemanticMembers() on the extension. The one exception is that CodingKeys is always synthesized in the type itself because you can define Encodable and Decodable conformances in different extensions and we want them to share the same CodingKeys, but we handle that separately below.

DougGregor replied:
It should be, unless those conformances are going back and inserting declarations into the main nominal type declaration. (like apparently `CodingKeys` does). Perhaps I need to do an audit of all of the conformance-synthesis code...

DougGregor replied:
Okay, I did a quick audit. `CodingKeys` is weird but handled specifically.

DougGregor committed:
[AST] Rename EmittedMembersRequest -> SemanticMembersRequest.

DougGregor committed:
[AST] Add IterableDeclContext::getParsedMembers().

Provide an accessor for retrieving the parsed members, generalizing
`ParseMembersRequest` so it can provide the parsed members for
deserialized/synthesized declarations as well. This is the counterpart
to the recently-generalized `getSemanticMembers()`; together, these
should suffice for most (all?) clients of `getMembers()`.

DougGregor commented:
@swift-ci smoke test

DougGregor commented:
@swift-ci please smoke test platform macOS

DougGregor commented:
@swift-ci please smoke test platform macOS

DougGregor commented:
@swift-ci please smoke test platform macOS

DougGregor commented:
@swift-ci please smoke test platform macOS

DougGregor commented:
macOS failure is unrelated (should be handled by https://github.com/apple/swift-driver/pull/235)
